sklearn/metrics/pairwise.py

```diff
@@ -673,7 +689,259 @@ def cosine_distances(X, Y=None):
+    the folowing aspects:
+
+    * The values in the original similarity S range between 0 and 1. To
+    guarantee this, it is assumed the numerical features of X and Y are
+    scaled between 0 and 1.
+
+    * Different from the original similarity S, this implementation
+    returns 1-S.
+    """
+    if issparse(X) or issparse(Y):
+        raise TypeError("Gower distance does not support sparse matrices")
+
+    if not isinstance(scale, (bool, list, np.ndarray)):
+        raise TypeError("Parameter scale must be boolean, list, or ndarray")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    y_is_not_none = Y is not None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    # as np.object, otherwise numeric columns will be converted to string
+    # if there are other string columns.
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X, dtype=np.object)
+
+    if y_is_not_none and not isinstance(Y, np.ndarray):
+        Y = np.asarray(Y, dtype=np.object)
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=X.dtype,
+                                 force_all_finite=False)
+
+    n_rows, n_cols = X.shape
+
+    categorical_features_obj, categorical_features_num, numerical_features = \
+        _detect_categorical_features(X, categorical_features)
+
+    # Categorical columns
+    X_cat_obj = X[:, categorical_features_obj].astype(np.object)
+    X_cat_num = X[:, categorical_features_num].astype(np.float)
+
+    # Numerical columns
+    X_num = X[:, numerical_features].astype(np.float)
+
+    Y_cat_obj = None
+    Y_cat_num = None
+    Y_num = None
+    if y_is_not_none:
+        Y_cat_obj = Y[:, categorical_features_obj].astype(np.object)
+        Y_cat_num = Y[:, categorical_features_num].astype(np.float)
+        Y_num = Y[:, numerical_features].astype(np.float)
+    else:
+        Y_cat_obj = X_cat_obj
+        Y_cat_num = X_cat_num
+        Y_num = X_num
+
+    num_cols = X_num.shape[1]
+    n_col_num_present = num_cols > 0
+    n_col_cat_obj_present = X_cat_obj.shape[1] > 0
+    n_col_cat_num_present = X_cat_num.shape[1] > 0
+
+    # Calculates the min and max values, and if requested, scale the
+    # input values in order to obtain the distances between 0 and 1,
+    # as proposed by the Gower's paper.
+    if n_col_num_present:
+        process_scale = False
+        if isinstance(scale, bool):
+            process_scale = scale
+        else:
+            if (isinstance(scale, list) and len(scale) != X_num.shape[1]) or\
+               (isinstance(scale, np.ndarray) and
+                    len(scale.flat) != X_num.shape[1]):
+                raise ValueError("Length of scale parameter must be equal "
+                                 "to the number of numerical columns.")
+            process_scale = True
+        # This needs to be called here to obtain the min and max,
```

Comment:
Alternatively, my preferred choice, would be to have a function `_check_gower_params(X, Y, scale, min, max)` or whatever which is also called within `_precompute_metric_params`, rather than rely so much on implementation details here.

Reply:
I prefer to simplify the _precompute_metric_params to calculate only the ranges. We don't need min and max as parameters, as long scale contains the ranges of of the entire data, the min and max of a sliced portion of that data still will be representative, this is fully supported by the tests between the lines 1086 and 1095 from test_pairwise.py.  I'll push this simplified version of _precompute_metric_params that I have in mind, then you can decide.